silx/image/bilinear.pyx

```diff
@@ -100,20 +145,76 @@ cdef class BilinearImage:
+                m1 = self.mask[i0, j1]
+                if m0 and m1:
+                    res = NAN
+                elif m0:
+                    res = self.data[i0, j1]
+                elif m1:
+                    res = self.data[i0, j0]
+                else:
+                    res = (self.data[i0, j0] * (y1 - d1)) + (self.data[i0, j1] * (d1 - y0))                
+            else:
+                res = (self.data[i0, j0] * (y1 - d1)) + (self.data[i0, j1] * (d1 - y0))
         elif j0 == j1:
-            res = (self.data[i0, j0] * (x1 - d0)) + (self.data[i1, j0] * (d0 - x0))
+            if self.has_mask:
+                m0 = self.mask[i0, j0]
+                m1 = self.mask[i1, j0]
+                if m0 and m1:
+                    res = NAN
+                elif m0:
+                    res = self.data[i1, j0]
+                elif m1:
+                    res = self.data[i0, j0]
+                else:
+                    res = (self.data[i0, j0] * (x1 - d0)) + (self.data[i1, j0] * (d0 - x0))                
+            else:
+                res = (self.data[i0, j0] * (x1 - d0)) + (self.data[i1, j0] * (d0 - x0))
         else:
-            res = (self.data[i0, j0] * (x1 - d0) * (y1 - d1))  \
-                + (self.data[i1, j0] * (d0 - x0) * (y1 - d1))  \
-                + (self.data[i0, j1] * (x1 - d0) * (d1 - y0))  \
-                + (self.data[i1, j1] * (d0 - x0) * (d1 - y0))
+            if self.has_mask:
+                m0 = self.mask[i0, j0]
+                m1 = self.mask[i1, j0]
+                m2 = self.mask[i0, j1]
+                m3 = self.mask[i1, j1]
+                if m0 and m1 and m2 and m3:
+                    res = NAN
+                else:
+                    m0 = not m0
+                    m1 = not m1
+                    m2 = not m2
+                    m3 = not m3
+                    if m0 and m1 and m2 and m3:
+                        res = (self.data[i0, j0] * (x1 - d0) * (y1 - d1))  \
+                            + (self.data[i1, j0] * (d0 - x0) * (y1 - d1))  \
+                            + (self.data[i0, j1] * (x1 - d0) * (d1 - y0))  \
+                            + (self.data[i1, j1] * (d0 - x0) * (d1 - y0))                    
+                    else:
+                        res = (m0 * self.data[i0, j0] * (x1 - d0) * (y1 - d1))  \
+                            + (m1 * self.data[i1, j0] * (d0 - x0) * (y1 - d1))  \
+                            + (m2 * self.data[i0, j1] * (x1 - d0) * (d1 - y0))  \
+                            + (m3 * self.data[i1, j1] * (d0 - x0) * (d1 - y0))
+                        scale = ((m0 * (x1 - d0) * (y1 - d1)) 
+                               + (m1 * (d0 - x0) * (y1 - d1))
+                               + (m2 * (x1 - d0) * (d1 - y0))
+                               + (m3 * (d0 - x0) * (d1 - y0)))/ \
+                                ((x1 - d0) * (y1 - d1) + (d0 - x0) * (y1 - d1)+
+                                 (x1 - d0) * (d1 - y0) + (d0 - x0) * (d1 - y0))
```

Comment:
If none of the 4 pixels underneath are masked, the scale is 1. This is the reason for the test just above.
If some pixels are masked, it gives a result which tries to be "meaningful" without performing a triangulation which would be the "ultimate" answer.

Reply:
I understood that you scale by the all the weights, but the denominator part of `scale`: `((x1 - d0) * (y1 - d1) + (d0 - x0) * (y1 - d1) + (x1 - d0) * (d1 - y0) + (d0 - x0) * (d1 - y0))` simplifies to 1 since:
```
        x0 = floor(d0)
        x1 = ceil(d0)
        y0 = floor(d1)
        y1 = ceil(d1)
```